pkg/collector/corechecks/snmp/snmp.go

```diff
@@ -0,0 +1,141 @@
+package snmp
+
+import (
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"time"
+)
+
+const (
+	snmpCheckName = "snmp"
+)
+
+// Check aggregates metrics from one Check instance
+type Check struct {
+	core.CheckBase
+	config  snmpConfig
+	session sessionAPI
+	sender  metricSender
+}
+
+// Run executes the check
+func (c *Check) Run() error {
+	start := time.Now()
+
+	sender, err := aggregator.GetSender(c.ID())
+	if err != nil {
+		return err
+	}
+	c.sender = metricSender{sender: sender}
+
+	staticTags := c.config.getStaticTags()
+
+	var checkErr error
+	tags, checkErr := c.doRun(staticTags)
+	if checkErr != nil {
+		c.sender.serviceCheck("snmp.can_check", metrics.ServiceCheckCritical, "", tags, checkErr.Error())
+	} else {
+		c.sender.serviceCheck("snmp.can_check", metrics.ServiceCheckOK, "", tags, "")
+	}
+
+	c.sender.gauge("snmp.devices_monitored", float64(1), "", tags)
+
+	// SNMP Performance metrics
+	c.sender.monotonicCount("datadog.snmp.check_interval", time.Duration(start.UnixNano()).Seconds(), "", tags)
+	c.sender.gauge("datadog.snmp.check_duration", time.Since(start).Seconds(), "", tags)
+	c.sender.gauge("datadog.snmp.submitted_metrics", float64(c.sender.submittedMetrics), "", tags)
+
+	// Commit
+	sender.Commit()
+	return checkErr
+}
+
+func (c *Check) doRun(staticTags []string) ([]string, error) {
+	tags := copyStrings(staticTags)
+
+	// Create connection
+	connErr := c.session.Connect()
```

Comment:
How does the library handle sessions? Do you need to call `Connect()` every time? Do we know if we can create and connect to a session once and then maybe the lib would handle reconnections automagically if they go stale? I vaguely recall from my SNMP days often the session could be setup just once and then reused ad-nauseam.  

Obviously the word "connection" has to be taken in the context of SNMP, as UDP is stateless. I'm just wondering how long those SNMP sesions can stick around for.

Reply:
My understanding is that [GoSNMP.Connect()](https://github.com/gosnmp/gosnmp/blob/master/gosnmp.go#L254-L262) used in the context of UDP calls will lead to the opening of a socket and `GoSNMP.Conn.Close()` will close that socket. By digging deeper, that's indeed what's happening under the hood:

```
func (sd *sysDialer) dialUDP(ctx context.Context, laddr, raddr *UDPAddr) (*UDPConn, error) {
	fd, err := internetSocket(ctx, sd.network, laddr, raddr, syscall.SOCK_DGRAM, 0, "dial", sd.Dialer.Control)
	if err != nil {
		return nil, err
	}
	return newUDPConn(fd), nil
}
```
source: udpsock_posix.go

```
// Close closes the connection.
func (c *conn) Close() error {
	if !c.ok() {
		return syscall.EINVAL
	}
	err := c.fd.Close()
	if err != nil {
		err = &OpError{Op: "close", Net: c.fd.net, Source: c.fd.laddr, Addr: c.fd.raddr, Err: err}
	}
	return err
}
```
source: net.go

Keeping the session open and reusing it would mean keeping the socket open for the lifetime of the check instance.

It seems reasonable to open socket and close it at each check run. What do you think ?